### PR TITLE
make aggregations deterministic

### LINKED ISSUE #9450: PromQL is not deterministic
Was trying to debug why a Grafana Heatmap panel for a histogram that clearly only had recorded observations in the 1M-2M and 2M-5M ranges looked like this:

Screenshot from 2021-10-05 12-53-51

You'll notice that the 2 buckets above had on the order of 30K observations each, whereas the other non-zero ones all show up as `0.0` (rounded). I also noticed that on refresh the `0.0` buckets shift around randomly. The underlying query is nothing fancy:

```
sum by (le) (
  avg by(le, some_label) (
    increase(metric[1m])
  )
)
```

Long story short, I looked at a specific timestamp and played around with the query and what seems to happen is that aggregations (such as `sum` and `avg`) produce randomly ordered results. If those results are in turn aggregated again, the result of that aggregation may be different depending on the order of the inputs.

Take for example the following query:

```
sum(
  avg by(foo) (
      label_replace(up{job="prometheus"}, "foo", "1", "job", "prometheus") * 0.99999999999998
    or
      label_replace(up{job="prometheus"}, "foo", "2", "job", "prometheus") * 100
    or
      label_replace(up{job="prometheus"}, "foo", "3", "job", "prometheus") * 100
  )
)
```

On any single-Prometheus deployment, this will sometimes yield `200.99999999999997`, sometimes `201`. The reason for that can be seen in [this playground example](https://play.golang.org/p/yZAGWt5OfiQ): `0.99999999999998 + 100.0 + 100.0 != 100.0 + 100.0 + 0.99999999999998`.

AFAICT the fix is to simply sort the output of `evaluator.aggregation()` by labels. Then in the example above `avg` will always produce the same output at a given timestamp. Not only that, but in the case of aggregating histogram buckets, the labels other than `le` will be sorted consistently across buckets, so the result of the aggregation will be exactly the same for buckets that have the same number of observations.

### PR DESCRIPTION
**Pull request by roidelapluie**
Fix #9450

### CODE REVIEW

**Comment by roidelapluie** [comment]
cc @bboreham for insights on the performance impact.

File: promql/engine.go
Comment (free) [inline promql/engine.go:2216]: Nice. This is actually a lot less overhead than explicitly sorting the results.

File: promql/testdata/aggregators.test
Comment (free) [inline promql/testdata/aggregators.test:509]: Also a lot more concise than my example.

Wondering if there's a way to make it a lot more likely for the test to catch non-determinism though. Best I can come up with is copy-pasting the `eval` statement a few times.

**Review by free — APPROVED** [review]
That was quick! And I like the fix a lot.

Thanks!

File: promql/testdata/aggregators.test
Comment (roidelapluie) [reply promql/testdata/aggregators.test:509]: It did fail 100% of the time on my machine.

Comment (roidelapluie) [reply promql/testdata/aggregators.test:509]: Note that every eval is evaluated multiple times: with an instant query, with a query range, with a @ timestamp.

**Comment by bboreham** [comment]
> cc @bboreham for insights on the performance impact.

Isn't that what benchmarks are for?

I am more concerned about the change to semantics; this should at least be documented.
Possibly it should be rejected - I may have misunderstood the issue this is fixing, but it seems to turn on below-precision differences in floating-point numbers, which should be expected to give surprising results.

**Comment by roidelapluie** [comment]
There is no change in semantics to me, PromQL should deterministic for a given order of inputs. This is a bug to me. I've claimed that in the past too, and I think it was after a discussion with other developers: https://github.com/prometheus/prometheus/issues/7400#issuecomment-655333558

**Comment by free** [comment]
Adding my 2c (not that anyone's asked for it): there is no change in semantics here, strictly speaking. The order of aggregation results was arbitrary and now it's deterministic (which one could argue is a certain kind of arbitrary).

Problem is that fundamentally it makes no sense for `x - x` to produce any other result than zero (across `x`'s label set). But with `x` an aggregation of aggregations (a quite common case, e.g. a sum of an average by `instance`) and without a deterministic ordering of aggregation results, this is not actually true in PromQL. In the same way that it makes no sense for valid (monotonically increasing) histogram buckets to become invalid after 2 rounds of aggregation.

**Comment by bboreham** [comment]
Forgive me if I misunderstood: “`sum by` returns results sorted by labels” seems a definite change in semantics to me. 

Even if you document this cannot be relied on, people will rely on it.

**Comment by roidelapluie** [comment]
> Forgive me if I misunderstood: “`sum by` returns results sorted by labels” seems a definite change in semantics to me.
> 
> Even if you document this cannot be relied on, people will rely on it.

We do not sort anything - we take things as they come and keep the order.

**Comment by bboreham** [comment]
BTW this is where they are sorted, right before the code changed in this PR gets called: https://github.com/prometheus/prometheus/blob/e261eccb35ee5ce4f024ab079c848e39660066ed/promql/engine.go#L1167-L1169

**Comment by roidelapluie** [comment]
I do not think the PromQL conformance tests run aggregations on top of aggregations, and this is where results could change.

**Comment by bboreham** [comment]
What I am worried about is that we can never make a future change that changes the order without breaking people that have come to rely on it. For comparison, Go maps intentionally randomise the order so people can't rely on it.

> TSDB also "happens to" return time series sorted by their labels

Is this true, or just something you think might be the case?  

The Prometheus maintainers could decide they don't care about this change.  That's fine, I just wanted to raise the issue and have it considered, not dismissed out of hand.

**Comment by prymitive** [comment]
> In Promql, we retain the order except in this case. It is an annoyance than you press execute multiple times in a console query you get different output.

I'd like to add that this is especially problematic with grafana and pie charts - I end up having to wrap all queries with `sort()`, otherwise metrics order is different on each refresh and so colours are assigned differently too, which causes visual noise.
It might be consider a problem with pie chart or a very minor issue, but IMHO it shows that based on approach taken by Prometheus here there might be consequence for downstream consumers of metrics - they might need to compensate for undefined order depending on use case. Having output more stable would make this whole class of issues (and extra code in all consumers) go away.

**Comment by bboreham** [comment]
> We are not bringing any ordering or sorting

Confused why you would repeat this.  I showed the exact line where sorting happens.

**Comment by codesome** [comment]
> We are not bringing any ordering or sorting - we just ensure the stability of aggregations for the same set of inputs. 

> We are sorting the labels, not the metrics,

These will matter for aggregation over aggregation, so bringing stability over current instability is indirectly bringing a requirement. But if we decide not to have have it as a requirement in compliance, then it should be fine I guess (but I don't see why we should not have this in compliance because we would like Prometheus compliant solutions to also have stability of aggregations, right?).

**Comment by free** [comment]
> Being sorted might just be a side effect of how storage is structured, but it not a requirement.

As Bryan pointed out above, if it's a behavior that people can rely on, they will rely on it and expect it to hold. It makes no difference whether it's intentional or an unintentional side-effect of the specific TSDB implementation.

If the fact that TSDB results "happen to be sorted" does not need to be documented, then neither does this. Or conversely, if this needs to be documented, then so does the fact that TSDB accidentally returns sorted results. (o:

As for aggregations sorting labels (not label values) lexicographically (which Bryan pointed to above), that is something that Prometheus already does now, not something introduced by this change. No idea whether that's documented anywhere, so possibly something else that should be explicitly documented, if you decide to go down that path.

**Comment by codesome** [comment]
I will be on PTO for more than a week now. So to not keep this hanging on me: I am not against this PR and I actually don't have a strong opinion on this for compliance, so I would go ahead with what you all decide is the best :), while my take is the stability should not be just in Prometheus but also mandated in compliant solutions (but that should not block this PR).

**Comment by roidelapluie** [comment]
it happens that we sort the select because we use mergequerier (we query heads+blocks, so it must be sorted to merge correctly)

We never use this shortcut:
https://github.com/prometheus/prometheus/blob/fdbc40a9efcc8197a94f23f0e479b0b56e52d424/storage/merge.go#L104

because we have 2 queriers all the time, which means we always want series sorted from TSDB:
https://github.com/prometheus/prometheus/blob/fdbc40a9efcc8197a94f23f0e479b0b56e52d424/tsdb/db.go#L1575


So, while PromQL uses Select(false), it is true in a Prometheus setup.

I'd just merge this PR as is, it seems generally good for our users to have reproducible results.

**Pull request merged**
